### PR TITLE
Close Channel when connect is cancelled during resolution

### PR DESCRIPTION
**Pull request by Gimini-3**
Motivation:

`Bootstrap.connect(...)` creates and registers a Channel before asynchronous address resolution. If the connect future is cancelled while resolution is pending, the Channel remains open until resolution completes. A late resolver failure also attempts to fail the cancelled promise and logs `IllegalStateException: complete already`.

Modification:

- Install the close-on-failure listener before address resolution starts so cancellation closes the registered Channel.
- Skip address resolution when cancellation wins the registration race.
- Ignore late resolver completion after the connect promise is done and use `tryFailure(...)` for cancellation races.
- Keep the resolver future active so shared in-flight resolutions are not cancelled.
- Add regression tests for cancellation during resolution with a real NIO Channel and before delayed registration completes.

Result:

Cancelling a Bootstrap connect closes its Channel while address resolution is pending, late resolver completion no longer tries to complete a cancelled promise, and delayed registration does not start unnecessary resolution.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Gimini-3 I think it would be better to call `setUncancellable(.....)` on the promise and don't use `try*`.

**Comment by Gimini-3** [comment]
@normanmaurer Thanks, updated. The promise is now marked uncancellable before address resolution starts, and the `try*` calls and late `isDone()` guard have been removed. Cancellation that wins before registration still closes the Channel and skips resolution. I also updated the tests to cover both boundaries; `./mvnw -pl transport test` passes (397 tests, 9 skipped).

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java [hunk 1]
```diff
@@ -240,7 +249,7 @@ private ChannelFuture doResolveAndConnect0(final Channel channel, SocketAddress
                 }
             });
         } catch (Throwable cause) {
-            promise.tryFailure(cause);
+            promise.setFailure(cause);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: this change should be reverted.

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java [hunk 2]
```diff
@@ -205,7 +213,8 @@ private ChannelFuture doResolveAndConnect0(final Channel channel, SocketAddress
                 resolver = ExternalAddressResolver.getOrDefault(externalResolver).getResolver(eventLoop);
             } catch (Throwable cause) {
                 channel.close();
-                return promise.setFailure(cause);
+                promise.setFailure(cause);
+                return promise;
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/bootstrap/Bootstrap.java:216]: just revert this change

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java (hunk 2 above)
Comment (Gimini-3) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java:216]: Reverted this to `return promise.setFailure(cause);`.

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java (hunk 1 above)
Comment (Gimini-3) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: Reverted this to `promise.tryFailure(cause);`.

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
@Gimini-3 hmm I think this needs some more thoughts... as the unit test failures show as with the current changes cancellation is not possible at all. So maybe your original changes were better. WDYT ?

**Comment by Gimini-3** [comment]
I agree that the original approach is better here. setUncancellable() changes the semantics of the entire connect operation and prevents cancellation even after address resolution, as SocketConnectionAttemptTest shows.  
The original approach preserves connect cancellation by installing CLOSE_ON_FAILURE before resolution, returning when the connect promise is already done, and safely ignoring late resolver completion without cancelling the resolver future. Would you prefer that I restore this approach?

**Comment by normanmaurer** [comment]
Yes

**Comment by Gimini-3** [comment]
Thanks, restored the original approach in b5f9f406fc. The relevant tests pass locally, and CI is running.

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java [hunk 3]
```diff
@@ -232,9 +240,12 @@ private ChannelFuture doResolveAndConnect0(final Channel channel, SocketAddress
 
             // Wait until the name resolution is finished.
             resolveFuture.addListener((FutureListener<SocketAddress>) future -> {
+                if (promise.isDone()) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: is this correct ? This means we would also not try to close the channel in this of a failure.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: transport/src/main/java/io/netty/bootstrap/Bootstrap.java (hunk 3 above)
Comment (Gimini-3) [reply transport/src/main/java/io/netty/bootstrap/Bootstrap.java]: Good point. Updated in `306143a1d7` to handle resolver failures first, so the Channel is always closed on failure. A completed promise now only skips `doConnect(...)` after successful resolution.

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
PR for 4.1 https://github.com/netty/netty/pull/17349
